src/runtime/rocm/rocm_device_api.cc

```diff
@@ -0,0 +1,142 @@
+    ROCM_CALL(hipSetDevice(ctx.device_id));
+    ROCM_CALL(hipFree(ptr));
+  }
+
+  void CopyDataFromTo(const void* from,
+                      size_t from_offset,
+                      void* to,
+                      size_t to_offset,
+                      size_t size,
+                      TVMContext ctx_from,
+                      TVMContext ctx_to,
+                      TVMStreamHandle stream) final {
+    hipStream_t hip_stream = static_cast<hipStream_t>(stream);
+    from = static_cast<const char*>(from) + from_offset;
+    to = static_cast<char*>(to) + to_offset;
+    if (ctx_from.device_type == kGPU && ctx_to.device_type == kGPU) {
+      ROCM_CALL(hipSetDevice(ctx_from.device_id));
+      if (ctx_from.device_id == ctx_to.device_id) {
+        GPUCopy(from, to, size, hipMemcpyDeviceToDevice, hip_stream);
+      } else {
+        hipMemcpyPeerAsync(to, ctx_to.device_id,
+                            from, ctx_from.device_id,
+                            size, hip_stream);
+      }
+    } else if (ctx_from.device_type == kGPU && ctx_to.device_type == kCPU) {
+      ROCM_CALL(hipSetDevice(ctx_from.device_id));
+      GPUCopy(from, to, size, hipMemcpyDeviceToHost, hip_stream);
+    } else if (ctx_from.device_type == kCPU && ctx_to.device_type == kGPU) {
+      ROCM_CALL(hipSetDevice(ctx_to.device_id));
+      GPUCopy(from, to, size, hipMemcpyHostToDevice, hip_stream);
+    } else {
+      LOG(FATAL) << "expect copy from/to GPU or between GPU";
+    }
+  }
+
+  void StreamSync(TVMContext ctx, TVMStreamHandle stream) final {
+    ROCM_CALL(hipSetDevice(ctx.device_id));
+    ROCM_CALL(hipStreamSynchronize(static_cast<hipStream_t>(stream)));
+  }
+
+  void SetStream(TVMContext ctx, TVMStreamHandle stream) final {
+    ROCMThreadEntry::ThreadLocal()
+        ->stream = static_cast<hipStream_t>(stream);
+  }
+
+  void* AllocWorkspace(TVMContext ctx, size_t size) final {
+    return ROCMThreadEntry::ThreadLocal()->pool.AllocWorkspace(ctx, size);
+  }
+
+  void FreeWorkspace(TVMContext ctx, void* data) final {
+    ROCMThreadEntry::ThreadLocal()->pool.FreeWorkspace(ctx, data);
+  }
+
+  static const std::shared_ptr<ROCMDeviceAPI>& Global() {
+    static std::shared_ptr<ROCMDeviceAPI> inst =
+        std::make_shared<ROCMDeviceAPI>();
+    return inst;
+  }
+
+ private:
+  static void GPUCopy(const void* from,
+                      void* to,
+                      size_t size,
+                      hipMemcpyKind kind,
+                      hipStream_t stream) {
+    if (stream != 0) {
+      ROCM_CALL(hipMemcpyAsync(to, from, size, kind, stream));
+    } else {
+      ROCM_CALL(hipMemcpy(to, from, size, kind));
+    }
+  }
+};
+
+typedef dmlc::ThreadLocalStore<ROCMThreadEntry> ROCMThreadStore;
+
+ROCMThreadEntry::ROCMThreadEntry()
+    : pool(kGPU, ROCMDeviceAPI::Global()) {
+}
+
+ROCMThreadEntry* ROCMThreadEntry::ThreadLocal() {
+  return ROCMThreadStore::Get();
+}
+
+TVM_REGISTER_GLOBAL("device_api.gpu")
```

Comment:
We will need to assign a new device type to rocm as gpu was reserved for CUDA already